reference-gen/pkg/generator/utils.go

```diff
@@ -0,0 +1,298 @@
+package generator
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+	"unicode"
+
+	"k8s.io/gengo/types"
+	"k8s.io/klog/v2"
+)
+
+var (
+	commonTypes = map[string]string{
+		"time.Duration": "duration",
+	}
+)
+
+// BEGIN: stringSet
+
+// stringSet is a utility for checking a set of strings
+type stringSet map[string]struct{}
+
+func (s stringSet) add(new string) {
+	s[new] = struct{}{}
+}
+
+func (s stringSet) has(required string) bool {
+	_, ok := s[required]
+	return ok
+}
+
+func (s stringSet) isEmpty() bool {
+	return len(s) == 0
+}
+
+func newStringSet(list []string) stringSet {
+	set := make(stringSet)
+	for _, s := range list {
+		set.add(s)
+	}
+	return set
+}
+
+// END: stringSet
+
+// BEGIN: typeSet
+
+// typeSet is a utility for checking a set of types
+type typeSet map[*types.Type]struct{}
+
+func (s typeSet) add(new *types.Type) {
+	s[new] = struct{}{}
+}
+
+func (s typeSet) has(required *types.Type) bool {
+	_, ok := s[required]
+	return ok
+}
+
+func (s typeSet) toList() []*types.Type {
```

Comment:
Optimization potential: build this up in a list object (in a struct that has both the set & list) as `add` is called so each later `toList` call isn't an O(N) operation.

Reply:
Provided we ensure that no duplicates end up in the list, that could work 🤔

Reply:
It would involve changing the type of the object to a struct though, I think I might leave this one for now